### DISCUSSION TITLE
Filter length for InceptionTime

### DISCUSSION THREAD

**Discussion by carbocation** [timeseriesAI/tsai#68] [Q&A] [answered]
In the InceptionTime manuscript, there is a plot of filter length vs accuracy, as a function of the length of the input time series:
![image](https://user-images.githubusercontent.com/218804/108803304-5d761380-7568-11eb-860b-2df02fda1aba.png)

I looked over the [source notebook](https://github.com/timeseriesAI/tsai/blob/master/nbs/102_models.InceptionTime.ipynb) for InceptionTime, but it wasn't clear to me how to change the filter length. Any pointers?

**Comment by vrodriguezf** [comment, marked as answer]
Is it not just `ks` for the first filter length of each `InceptionModule`? The rest are computed based on that one:
```python
ks = [ks // (2**i) for i in range(3)]
```

**Reply by oguiza** [reply]
Yes, exactly.  You can modify filter length by changing ks. The default is 40, which lead to filters of size [39, 19, 9].
There's another calculation that follows that one to ensure that filters only contain odd numbers as it doesn't make sense to have a kernel with an even number with a padding='same' convolution. 
```
ks = [ks // (2**i) for i in range(3)]
ks = [k if k % 2 != 0 else k - 1 for k in ks]
```

**Reply by carbocation** [reply]
Oh, that is great - thank you!

**Reply by dvirnimrod** [reply]
@oguiza can you please explain why doesn't it make sense to use an even sized kernel with a padding='same' convolution?
(as in the paper they didn't address this issue and used even numbers)

**Reply by oguiza** [reply]
Hi @dvirnimrod,

It's a great question.
It's because when you use padding, you want to use the same amount of padding on both extremes of the time series. This is only possible with an odd kernel_size. That's why you barely ever see even kernel sizes. 

In the paper, it's true they mention 40. But in the code, the way they calculate it is: 

```
self.kernel_size = kernel_size - 1
```
```
 # kernel_size_s = [3, 5, 8, 11, 17]
kernel_size_s = [self.kernel_size // (2 ** i) for i in range(3)]
```
So in the end, kernel_size_s = [39, 19, 9]
I hope this helps.

**Reply by dvirnimrod** [reply]
Thanks @oguiza !

I see your point.
Even though, for my understanding, the "same" padding allows one to have different padding on two extremes, if one wishes :)

As regards to the paper, in the official implementation they set kernel_size=41, so after the code lines you've posted here, in the end it's [40, 20, 10], as they declare in the paper itself. Am I missing here something?

**Reply by oguiza** [reply]
You are right, that you can have different padding on each extreme. But in my opinion, that doesn't make any sense as you are also using multiple layers. 

I can't explain why the author set the default kernel_size=41. In the main.py file, the use even kernel sizes. You would need to ask them. Sorry I can't help.

**Reply by dvirnimrod** [reply]
I've just wanted to make sure I'm not missing anything.
You've surely helped, thanks!
